### PR TITLE
Enhances AI law and borg logging

### PR DESCRIPTION
**Pull request by bloons3**
## About The Pull Request

Ports https://github.com/BeeStation/BeeStation-Hornet/pull/4521

> ADMIN LOG: bloons3/(Julio Jenner)(?) (FLW) used 'Quarantine' AI Module on @bloons3[DC]/(Z-3)(?) (FLW) from AI Upload Chamber (112, 129, 2). , , impacting synced borgs no key/(Security Cyborg-361)(?) (FLW),

image

image

## Why It's Good For The Game

Persistent logging for sometimes disputed actions
- The "impacted cyborgs" is useful for reconstructing which borgs were synced to an AI at what time, which sometimes may not be clear from logs

## Changelog
:cl:
admin: Adds logging for impacted cyborgs during law changes
admin: Adds additional cyborg logging
/:cl:

### CODE REVIEW

**Comment by Dorsisdwarf** [comment]
Whats with the full stop comma comma in your example message?

**Comment by bloons3** [comment]
> Whats with the full stop comma comma in your example message?

image

Was part of the previous message

**Comment by bloons3** [comment]
Now that I think about it, should `[user.key]/[user.name]` be replaced by `[ADMIN_LOOKUP(user)]`?

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 1]
```diff
@@ -581,15 +581,15 @@
...
 		smoke.start()
 		sleep(2)
 		for(var/i in 1 to 4)
 			playsound(R, pick('sound/items/drill_use.ogg', 'sound/items/jaws_cut.ogg', 'sound/items/jaws_pry.ogg', 'sound/items/welder.ogg', 'sound/items/ratchet.ogg'), 80, TRUE, -1)
 			sleep(12)
 		if(!prev_lockcharge)
-			R.SetLockdown(0)
-		R.set_anchored(FALSE)
+			R.SetLockdown(FALSE)
+		R.anchored = FALSE
```
Comment (Rohesie) [inline code/game/objects/items/robot/robot_upgrades.dm:592]: Is this change intentional, or the byproduct of a merge conflict?

File: code/game/objects/items/AI_modules.dm [hunk 2]
```diff
@@ -68,9 +69,22 @@ AI MODULES
...
+		var/mob/living/silicon/ai/owner = law_datum.owner
+		for(var/mob/living/silicon/robot/owned_borg in owner.connected_robots)
+			if(owned_borg.connected_ai && owned_borg.lawupdate)
+				affected_cyborgs += owned_borg
+				borg_flw += "[ADMIN_LOOKUPFLW(owned_borg)], "
+				borg_txt += "[ADMIN_LOOKUP(owned_borg)], "
+
+	GLOB.lawchanges.Add("[time] <B>:</B> [user.name]([user.key]) used [src.name] on [ainame]([aikey]).[law2log ? " The law specified [law2log]" : ""], [length(affected_cyborgs) ? ", impacting synced borgs [borg_txt]" : "FALSE"]")
+	log_law("[user.key]/[user.name] used [src.name] on [aikey]/([ainame]) from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""] , [length(affected_cyborgs) ? ", impacting synced borgs [borg_txt]" : "FALSE"]")
+	message_admins("[ADMIN_LOOKUPFLW(user)] used [src.name] on [ADMIN_LOOKUPFLW(law_datum.owner)] from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""] , [length(affected_cyborgs) ? ", impacting synced borgs [borg_flw]" : "FALSE"]")
```
Comment (Rohesie) [inline code/game/objects/items/AI_modules.dm:87]:
```suggestion
	var/list/borg_txt = list()
	var/list/borg_flw = list()
	if(isAI(law_datum.owner))
		var/mob/living/silicon/ai/owner = law_datum.owner
		for(var/mob/living/silicon/robot/owned_borg in owner.connected_robots)
			if(owned_borg.connected_ai && owned_borg.lawupdate)
				affected_cyborgs += owned_borg
				borg_flw += "[ADMIN_LOOKUPFLW(owned_borg)], "
				borg_txt += "[ADMIN_LOOKUP(owned_borg)], "

	borg_txt = borg_txt.Join()
	GLOB.lawchanges.Add("[time] <B>:</B> [user.name]([user.key]) used [src.name] on [ainame]([aikey]).[law2log ? " The law specified [law2log]" : ""], [length(affected_cyborgs) ? ", impacting synced borgs [borg_txt]" : "FALSE"]")
	log_law("[user.key]/[user.name] used [src.name] on [aikey]/([ainame]) from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""] , [length(affected_cyborgs) ? ", impacting synced borgs [borg_txt]" : "FALSE"]")
	message_admins("[ADMIN_LOOKUPFLW(user)] used [src.name] on [ADMIN_LOOKUPFLW(law_datum.owner)] from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""] , [length(affected_cyborgs) ? ", impacting synced borgs [borg_flw.Join()]" : "FALSE"]")
```
Prefer to use lists and then `.Join()` for accumulating text. They're more efficient than byond text appending.

**Comment by Rohesie** [comment]
> 
> 
> Now that I think about it, should `[user.key]/[user.name]` be replaced by `[key_name(user)]`?

If you think it gives better information, go ahead.

File: code/game/objects/items/robot/robot_upgrades.dm (hunk 1 above)
Comment (bloons3) [reply code/game/objects/items/robot/robot_upgrades.dm:592]: merge conflict, resolving

File: code/game/objects/items/robot/robot_upgrades.dm [hunk 3]
```diff
@@ -581,15 +581,15 @@
 		R.SetLockdown(TRUE)
 		R.set_anchored(TRUE)
 		var/datum/effect_system/smoke_spread/smoke = new
-		smoke.set_up(1, R.loc)
+		smoke.set_up(TRUE, R.loc)
```
Comment (bloons3) [inline code/game/objects/items/robot/robot_upgrades.dm:584]: this is 1 radius, not Truthy/Falsy

File: code/game/objects/items/AI_modules.dm [hunk 4]
```diff
@@ -68,9 +69,23 @@ AI MODULES
...
-	log_law("[user.key]/[user.name] used [src.name] on [aikey]/([ainame]) from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""]")
-	message_admins("[ADMIN_LOOKUPFLW(user)] used [src.name] on [ADMIN_LOOKUPFLW(law_datum.owner)] from [AREACOORD(user)].[law2log ? " The law specified [law2log]" : ""]")
+
+	//affected cyborgs are cyborgs linked to the AI with lawsync enabled
+	var/affected_cyborgs = list()
+	var/list/borg_txt = list()
+	var/list/borg_flw = list()
+	if(isAI(law_datum.owner))
+		var/mob/living/silicon/ai/owner = law_datum.owner
+		for(var/mob/living/silicon/robot/owned_borg in owner.connected_robots)
```
Comment (Rohesie) [inline code/game/objects/items/AI_modules.dm:79]:
```suggestion
		for(var/mob/living/silicon/robot/owned_borg as anything in owner.connected_robots)
```
Can there be a type other than this one in that list?
If not, then there's no need for a type check.

**Comment by Rohesie** [comment]
Should be ready for merge after this last change.
Set it up as ready for review when done.
